File: artifacts/api-server/server/routes/reading_plans.py
from fastapi import APIRouter, Depends

from ..database import db_cursor
from ..auth import require_auth

router = APIRouter()


@router.get("/reading-plans")
def list_reading_plans(user_id: int = Depends(require_auth)):
    with db_cursor() as cur:
        cur.execute("SELECT * FROM reading_plans")
        plans = [dict(r) for r in cur.fetchall()]
        cur.execute("SELECT * FROM reading_plan_days ORDER BY day_number ASC")
        days = [dict(r) for r in cur.fetchall()]

    plans_out = []
    for p in plans:
        plan_days = [
            {
                "id": d["id"],
                "planId": d["plan_id"],
                "dayNumber": d["day_number"],
                "passage": d["passage"],
                "reflection": d["reflection"],
            }
            for d in days if d["plan_id"] == p["id"]
        ]
        plans_out.append({
            "id": p["id"],
            "title": p["title"],
            "description": p["description"],
            "createdAt": p["created_at"],
            "days": plan_days,
        })
    return {"plans": plans_out}


@router.get("/reading-plans/progress")
def get_reading_progress(user_id: int = Depends(require_auth)):
    with db_cursor() as cur:
        cur.execute(
            "SELECT * FROM reading_progress WHERE user_id = %s",
            (user_id,),
        )
        rows = [dict(r) for r in cur.fetchall()]
    return {"progress": [{
        "id": r["id"],
        "userId": r["user_id"],
        "planDayId": r["plan_day_id"],
        "completed": r["completed"],
        "completedAt": r["completed_at"],
    } for r in rows]}


@router.post("/reading-plans/days/{day_id}/complete", status_code=201)
def complete_reading_day(day_id: int, user_id: int = Depends(require_auth)):
    with db_cursor() as cur:
        cur.execute(
            "SELECT * FROM reading_progress WHERE user_id = %s AND plan_day_id = %s LIMIT 1",
            (user_id, day_id),
        )
        existing = cur.fetchone()

    if existing:
        row = dict(existing)
        return {"progress": {
            "id": row["id"],
            "userId": row["user_id"],
            "planDayId": row["plan_day_id"],
            "completed": row["completed"],
            "completedAt": row["completed_at"],
        }}

    with db_cursor(commit=True) as cur:
        cur.execute(
            "INSERT INTO reading_progress (user_id, plan_day_id) VALUES (%s, %s) RETURNING *",
            (user_id, day_id),
        )
        row = dict(cur.fetchone())
    return {"progress": {
        "id": row["id"],
        "userId": row["user_id"],
        "planDayId": row["plan_day_id"],
        "completed": row["completed"],
        "completedAt": row["completed_at"],
    }}

File: artifacts/api-server/src/routes/reading-plans.ts
import { Router, type IRouter } from "express";
import { eq, and } from "drizzle-orm";
import {
  db,
  readingPlansTable,
  readingPlanDaysTable,
  readingProgressTable,
} from "@workspace/db";
import { requireAuth, type AuthedRequest } from "../lib/auth";

const router: IRouter = Router();

router.get("/reading-plans", requireAuth, async (_req, res) => {
  const plans = await db.select().from(readingPlansTable);
  const days = await db.select().from(readingPlanDaysTable);
  return res.json({
    plans: plans.map((p) => ({
      ...p,
      days: days
        .filter((d) => d.planId === p.id)
        .sort((a, b) => a.dayNumber - b.dayNumber),
    })),
  });
});

router.get("/reading-plans/progress", requireAuth, async (req: AuthedRequest, res) => {
  const rows = await db
    .select()
    .from(readingProgressTable)
    .where(eq(readingProgressTable.userId, req.userId!));
  return res.json({ progress: rows });
});

router.post("/reading-plans/days/:dayId/complete", requireAuth, async (req: AuthedRequest, res) => {
  const planDayId = Number(req.params.dayId);
  const existing = await db
    .select()
    .from(readingProgressTable)
    .where(
      and(
        eq(readingProgressTable.userId, req.userId!),
        eq(readingProgressTable.planDayId, planDayId),
      ),
    )
    .limit(1);
  if (existing.length > 0) {
    return res.json({ progress: existing[0] });
  }
  const [row] = await db
    .insert(readingProgressTable)
    .values({ userId: req.userId!, planDayId })
    .returning();
  return res.status(201).json({ progress: row });
});

export default router;
